### ISSUE TITLE
Performance: init ... endinit block for explicit engine

### ISSUE THREAD

**Issue by kleinj** [prismmodelchecker/prism#2]
When using the `init ... endinit` method of specifying multiple initial states, the state space generation in the explicit engine can be quite slow. Currently, all possible valuations of the state variables are generated, i.e., the full Cartesian product, and each of them is checked against the init expression to determine whether it is an actual initial state (`simulator.ModulesFileModelGenerator.getInitialStates()`). If the potential state space is quite large, this becomes very expensive, even though potentially only a small fragment of the state space is actually reachable. Clearly, it is a hard problem for the explicit engine to generally find all the satisfying variable assignments, but it would be nice if the non-hard cases would be handled more efficiently.

Some ideas for possible approaches:
 * Detect common, "simple" init expression patterns (e.g., conjunction of simple comparison expressions) and handle those separately.

* Exploit the symbolic MTBDD engine by translating only the init expression into a BDD and then iterate over the satisfying variable assignments to obtain the initial states.

* Generating the satisfying assigments via some kind of structural induction? There, it would probably be good to transform the expression into positive normal form, i.e., with no more boolean negation operator.

**Comment by kleinj** [comment]
I guess it would be nice to detect at least if the expression is in disjunctive normal form, i.e., a disjunction of conjunctive clauses, and with only equality comparisons. There, it would be rather straightforward to find the satisfying assignments and it would cover the common use case where one wants to have more than a single initial state but the number of initial states is rather small, and there is no other way to specify this without using an init expression.

**Comment by davexparker** [comment]
Yes, this would be nice. And I think there plenty of simple common cases (`true`, conjunction of inequalities) that could be handled quite easily.

For handling the general case, an alternative is to push the work to a SAT solver (in ALLSAT mode) but that might be overkill except for complicated expressions with smallish number of satisfying valuations.

**Comment by merkste** [comment]
As we ran into that issue as well recently, I implemented a slightly more efficient generator-like approach. Instead of building a list of all possible states and then filtering that list, it enumerates the cartesian product of the variable domains and filters them on the fly. It should be straightforward to extend to support the basic patterns as mentioned by @kleinj.

**Comment by davexparker** [comment]
Thanks @merkste. Code/patches welcome.

**Comment by merkste** [comment]
As there is a second solution, using the mtbbd engine on the horizon, I'd like to suggest to integrate both approaches and make an option. The rationale behind this is that there are some models that just do not work with the mtbdd engine and so do their initial conditions. What do you think?
